api/mysagw/oidc_auth/tests/test_authentication.py

```diff
@@ -0,0 +1,102 @@
+import hashlib
+import json
+
+import pytest
+from django.core.cache import cache
+from mozilla_django_oidc.contrib.drf import OIDCAuthentication
+from requests.exceptions import HTTPError
+from rest_framework import exceptions, status
+from rest_framework.exceptions import AuthenticationFailed
+
+from mysagw.identity.models import Identity
+
+
+@pytest.mark.parametrize("is_id_token", [True, False])
+@pytest.mark.parametrize(
+    "authentication_header,authenticated,error",
+    [
+        ("", False, False),
+        ("Bearer", False, True),
+        ("Bearer Too many params", False, True),
+        ("Basic Auth", False, True),
+        ("Bearer Token", True, False),
+    ],
+)
+def test_authentication(
+    db,
+    rf,
+    authentication_header,
+    authenticated,
+    error,
+    is_id_token,
+    requests_mock,
+    settings,
+):
+    userinfo = {"sub": "1"}
+    requests_mock.get(settings.OIDC_OP_USER_ENDPOINT, text=json.dumps(userinfo))
+
+    assert Identity.objects.count() == 0
+
+    if not is_id_token:
+        userinfo = {"client_id": "test_client", "sub": "1"}
+        requests_mock.get(
+            settings.OIDC_OP_USER_ENDPOINT, status_code=status.HTTP_401_UNAUTHORIZED
+        )
+        requests_mock.post(
+            settings.OIDC_OP_INTROSPECT_ENDPOINT, text=json.dumps(userinfo)
+        )
+
+    request = rf.get("/openid", HTTP_AUTHORIZATION=authentication_header)
+    try:
+        result = OIDCAuthentication().authenticate(request)
+    except exceptions.AuthenticationFailed:
+        assert error
+    else:
+        if result:
```

Comment:
Instead of branching on the authentication result, I'd branch on the parametrisation instead. That way, you test the assumptions and are not "running after" the code being tested.

Reply:
Done